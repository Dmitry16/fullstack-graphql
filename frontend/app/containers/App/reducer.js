/*
 * AppReducer
 */

import { fromJS } from 'immutable';

import {
  LOAD_MOVIES_SUCCESS,
  LOAD_MOVIES,
  LOAD_MOVIES_ERROR,
} from './constants';

// The initial state of the App
const initialState = fromJS({
  loading: false,
  error: false,
  currentMovieTitle: false,
  movieData: {
    movies: false,
  },
});

function appReducer(state = initialState, action) {
  switch (action.type) {
    case LOAD_MOVIES:
      return state
        .set('loading', true)
        .set('error', false)
        .setIn(['movieData', 'movies'], false)
    case LOAD_MOVIES_SUCCESS:
      return state
        .setIn(['movieData', 'movies'], action.movies)
        .set('loading', false)
        .set('currentMovieTitle', action.currentMovieTitle);
    case LOAD_MOVIES_ERROR:
      return state
        .set('error', action.error)
        .set('loading', false);
    default:
      return state;
  }
}

export default appReducer;
