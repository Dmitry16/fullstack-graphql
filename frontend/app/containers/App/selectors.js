/**
 * The global state selectors
 */

import { createSelector } from 'reselect';

const selectGlobal = (state) => state.get('global');

const selectRoute = (state) => state.get('route');

const makeSelectCurrentMovie = () => createSelector(
  selectGlobal,
  (globalState) => globalState.get('currentMovieTitle')
);

const makeSelectLoading = () => createSelector(
  selectGlobal,
  (globalState) => globalState.get('loading')
);

const makeSelectError = () => createSelector(
  selectGlobal,
  (globalState) => globalState.get('error')
);

const makeSelectMovies = () => createSelector(
  selectGlobal,
  (globalState) => globalState.getIn(['movieData', 'movies'])
);

const makeSelectLocation = () => createSelector(
  selectRoute,
  (routeState) => routeState.get('location').toJS()
);

export {
  selectGlobal,
  makeSelectCurrentMovie,
  makeSelectLoading,
  makeSelectError,
  makeSelectMovies,
  makeSelectLocation,
};
