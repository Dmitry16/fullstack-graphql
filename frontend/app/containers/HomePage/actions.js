/*
 * Home Actions
 *
 * Actions change things in your application
 * Since this boilerplate uses a uni-directional data flow, specifically redux,
 * we have these actions which are the only way your application interacts with
 * your application state. This guarantees that your state is up to date and nobody
 * messes it up weirdly somewhere.
 *
 * To add a new Action:
 * 1) Import your constant
 * 2) Add a function like this:
 *    export function yourAction(var) {
 *        return { type: YOUR_ACTION_CONSTANT, var: var }
 *    }
 */

import { 
  CHANGE_MOVIE_TITLE,
  CHECK_SESSION_STORAGE,
  SET_SESSION_STORAGE
} from './constants';

/**
 * Changes the input field of the form
 *
 * @param  {movieTitle} The new text of the input field
 *
 * @return {object}    An action object with a type of CHANGE_USERNAME
 */
export function changeMovieTitle(movieTitle) {
  return {
    type: CHANGE_MOVIE_TITLE,
    movieTitle
  };
}

export function setSessionStorage(name, val) {
  sessionStorage.setItem(name, JSON.stringify(val));
  return {
    type: SET_SESSION_STORAGE,
    name,
    val
  };
}
