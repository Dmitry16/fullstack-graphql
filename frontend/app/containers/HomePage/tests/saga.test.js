/**
 * Tests for HomePage sagas
 */

import { put, takeLatest } from 'redux-saga/effects';

import { LOAD_MOVIES } from 'containers/App/constants';
import { moviesLoaded, movieLoadingError } from 'containers/App/actions';

import moviesData, { getMovies } from '../saga';

const currentMovieTitle = 'Blade Runner';

/* eslint-disable redux-saga/yield-effects */
describe('getMovies Saga', () => {
  let getMoviesGenerator;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    getMoviesGenerator = getMovies();

    const selectDescriptor = getMoviesGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor = getMoviesGenerator.next(currentMovieTitle).value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should call the movieLoadingError action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = getMoviesGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(movieLoadingError(response)));
  });
});

describe('moviesData Saga', () => {
  const moviesDataSaga = moviesData();

  it('should start task to watch for LOAD_REPOS action', () => {
    const takeLatestDescriptor = moviesDataSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(LOAD_MOVIES, getMovies));
  });
});
