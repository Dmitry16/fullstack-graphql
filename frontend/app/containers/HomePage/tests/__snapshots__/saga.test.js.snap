// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getMovies Saga should call the movieLoadingError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`getMovies Saga should call the movieLoadingError action if the response errors 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "http://localhost:3003/api/search?keyword=Blade Runner",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`getMovies Saga should dispatch the moviesLoaded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`getMovies Saga should dispatch the moviesLoaded action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "http://localhost:3003/api/search?keyword=Blade Runner",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
